### PR TITLE
Add bulk directory support to homed

### PR DESCRIPTION
**Pull request by AdrianVovk**
This defines some storage space that gets automatically copied out of the home directory and into some shared system space (i.e. /var/cache). Thus, the contents of this folder remain available even if the homed user is not currently activated

This provides for some storage space that can be used to store non-private user data that still conceptually belongs to the home directory.  This may be used for any number of things, but most immediately:

- Storing the user's profile picture
- Storing a blurred version of the user's wallpaper (to be displayed by GDM)

These are large blobs of data that cannot be stored in the user record itself

### CODE REVIEW

**Comment by AdrianVovk** [comment]
Ok I tested this all and it works. I think it's ready for a look.

File: src/basic/time-util.h [hunk 1]
```diff
@@ -112,10 +113,15 @@ static inline bool triple_timestamp_is_set(const triple_timestamp *ts) {
 usec_t triple_timestamp_by_clock(triple_timestamp *ts, clockid_t clock);
 
 usec_t timespec_load(const struct timespec *ts) _pure_;
+_pure_ static inline usec_t statx_timestamp_load(const struct statx_timestamp *ts) {
+        return timespec_load(&(const struct timespec) { .tv_sec = ts->tv_sec, .tv_nsec = ts->tv_nsec });
+}
 nsec_t timespec_load_nsec(const struct timespec *ts) _pure_;
+_pure_ static inline nsec_t statx_timestamp_load_nsec(const struct statx_timestamp *ts) {
```
Comment (poettering) [inline src/basic/time-util.h]: i'd drop the `_pure_`, the compiler should always be able to figure that out on its own here. 

the `_pure_` stuf makes sense for functions whose implementation is hidden away in .c files, where the compiler cannot look into (because it becomes available only during link time). But for static inline stuff defined in headers this is unnecessary, the compiler can always make the determinations on its own, because it has the full sources around.

File: src/basic/time-util.h [hunk 2]
```diff
@@ -112,10 +113,15 @@ static inline bool triple_timestamp_is_set(const triple_timestamp *ts) {
...
+        return timespec_load_nsec(&(const struct timespec) { .tv_sec = ts->tv_sec, .tv_nsec = ts->tv_nsec });
+}
```
Comment (poettering) [inline src/basic/time-util.h]: these helpers make a ton of sense, but I think they should be defined in stat-util.h, somewhere close to statx_fallback() or STRUCT_NEW_STATX_DEFINE()...

After all these are more helpers for statx() than for timespec.

File: src/shared/copy.c [hunk 3]
```diff
@@ -1170,8 +1170,24 @@ static int fd_copy_tree_generic(
...
+                        struct stat buf;
+
+                        replace_ts = timespec_load_nsec(&st->st_mtim);
+                        if (fstatat(dt, to, &buf, AT_SYMLINK_NOFOLLOW) >= 0)
+                                orig_ts = timespec_load_nsec(&buf.st_mtim);
+
+                        if (orig_ts == NSEC_INFINITY || replace_ts == NSEC_INFINITY)
+                                return r; /* If we couldn't get both modification times, play it safe and don't overwrite. */
+
+                        if (orig_ts >= replace_ts)
```
Comment (poettering) [inline src/shared/copy.c:1187]: hmm, why check for newer? why not check for different?

I mean, I think we should not expect that if a home dir is portable between systems that all systems have correct clocks.

I think it would be more robust to just check for different timestamps here, rather than newer/older. 

Or in other words, use something like stat_inode_unmodified().

File: src/shared/copy.c [hunk 4]
```diff
@@ -1455,6 +1471,26 @@ int copy_file_atomic_at_full(
...
+                if (fstatat(dir_fdf, from, &buf, AT_SYMLINK_NOFOLLOW) < 0)
+                        return -errno;
+                replace_ts = timespec_load_nsec(&buf.st_mtim);
+
+                if (fstatat(dir_fdt, to, &buf, AT_SYMLINK_NOFOLLOW) < 0)
+                        return -errno;
+                orig_ts = timespec_load_nsec(&buf.st_mtim);
+
+                assert(orig_ts != NSEC_INFINITY);
+                assert(replace_ts != NSEC_INFINITY);
```
Comment (poettering) [inline src/shared/copy.c:1487]: assertions for this? I don't think we can assume that the timestamps set are always valid. for example fuse sets the timestamps to -1 iirc in some cases. I think we must be prepared here for totally invalid timestamps.

hence, here too, stat_inode_unmodified() seems like the way to go

File: man/systemd-homed-sync-bulk.service.xml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  Copyright © 2023 GNOME Foundation Inc.
+        Original Author: Adrian Vovk
+-->
+
+<refentry id="systemd-homed-sync-bulk.service" conditional='ENABLE_HOMED'
+          xmlns:xi="http://www.w3.org/2001/XInclude">
+
+  <refentryinfo>
+    <title>systemd-homed-sync-bulk.service</title>
```
Comment (poettering) [inline man/systemd-homed-sync-bulk.service.xml:16]: hmm, why make this a service of its own? I would just put this inside of homework.c i.e. as part of the reconciliation of the user record in the home dir and on the host. I mean, we are not talking about copying gigabytes here, but just a couple of 10M or so.

File: man/systemd-homed-sync-bulk.service.xml [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+    <ulink url="https://systemd.io/USER_RECORD/">JSON User Record</ulink>), but is too large to embed directly
+    into the JSON. For example: the bulk directory should be used to store a custom profile picture for the user.
+    Another example: the bulk directory could be used to store the user's wallpaper, which will then allow the
+    system's login UI to display it.</para>
+
+    <para>As mentioned above, there are two copies of the bulk directory that are kept in sync. The first is
+    <filename>$XDG_DATA_HOME/homed/</filename> or <filename>$HOME/.local/share/homed/</filename>, which we will
+    call the "home bulk" directory. The home bulk directory is only accessible while the user is in an active
+    state (i.e. logged in and unlocked). The other copy is found in
+    <filename>/var/cache/homed/<replaceable>username</replaceable>/</filename>,
```
Comment (poettering) [inline man/systemd-homed-sync-bulk.service.xml:53]: hmm, I think this is systemd-specific enough to warrant to be placed inside of /var/cache/systemd/homed/, i.e. to put the "systemd" name in there.

usually, we only put super generic stuff we think makes sense for alternative implementsions too, outside of "systemd" dirs in /usr/lib/, /var or /run

**Comment by poettering** [comment]
hmm, so this code runs as user, but how does it get write access to the dir in /var/ then?

I am pretty sure this should be implemented in privileged code, and enforce some super restrictive rules, i.e:

1. only copy regular files, no other inode types
2. do not descend into any dir tree
3. source should probably be placed in the top-level dir of $HOME, rather then XDG subdirs, to avoid symlink vulnerabilities.
4. as we copy, sum up sizes, and refuse copying more than some hardcoded limit (i.e. 64M or so)
5. do not copy access mode, ownership, xattrs or such. force a fixed access mode on copied files, i.e. no x bit, no suid, world readable and stuff
6. be very restrictive with the the filename charset we accept. i.e. no control chars, maybe not even non-ascii.

**Comment by poettering** [comment]
conceptually looks good and makes a ton of sense to me.

File: src/home/homed-home.c [hunk 7]
```diff
@@ -334,6 +342,11 @@ int home_unlink_record(Home *h) {
         if (unlink(fn) < 0 && errno != ENOENT)
                 return -errno;
 
+        fn = strjoina(home_system_bulk_dir(), "/", h->user_name);
```
Comment (poettering) [inline src/home/homed-home.c:345]: path_join() please.

File: docs/USER_RECORD.md [hunk 8]
```diff
@@ -248,6 +252,13 @@ sessions.
 purpose of an avatar. This must be a string, and should follow the semantics
 defined in the [Icon Naming
 Specification](https://standards.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.html).
+You should probably use `avatarPath` instead.
+
+`avatarPath` → The path to a file containing the user's avatar image, a string.
+If the path is relative, it will be resolved relative to the user's bulk
+directory. The file referenced by this setting should be world-readable
+(i.e. should have a mode of 0644). If both `avatarPath` and `iconName` are
+set, `avatarPath` should take priority.
```
Comment (poettering) [inline docs/USER_RECORD.md]: hmm, do we really want to do this like this? given that images are supposed to be portable i think we should not allow absolute paths here at all, because they only make sense relative to some system.

And we actually want that any user management system copies the avatar picture into the home dir, so it is truly portable, so we really want to take away the ability to reference external reosurces.

But I wonder, should this really be a json property at all? i.e. do we need the indirection? because we could just say the avatar is placed in a file "avatar-image" or so in the user's bulk dir. (without any suffix, so that PNG, JPEG, whatever is actually fine, parsers would have to deal with that).

i mean, it appears to me that this bulk dir mostly contains "blobs", i.e. data that is relatively unstructured from homed's pov, hence it should bs sufficient to give the files generic identifiers via filenames and that's all.

or in other words, instead of adding a new json property here to the spec, add a new section to the bottom, which lists "well-known" file names in the bulk dir, the avatar being the most prominent of them.

File: shell-completion/bash/homectl [hunk 9]
```diff
@@ -50,7 +50,7 @@ _homectl() {
                         --realm
                         --email-address
                         --location
-                        --icon-name
+                        --avatar-path
```
Comment (poettering) [inline shell-completion/bash/homectl]: why drop the icon name?

**Comment by poettering** [comment]
> hmm, so this code runs as user, but how does it get write access to the dir in /var/ then?

ignore this question, i found this now.

**Comment by poettering** [comment]
Somehow i missed the last two commits when I looked at this earlier.

Anyway, this looks good, but I think it's important this data is copied from priv code, outside of control of the user. After all, this is supposed to be world-accessible, hence we want some basic sanitizing of the data, so that nothing "weird" is copied, just regular, safely named files. Also, we it's problematic if regular users get write access to a new dir in /var/ that typically has no quota assigned. Right now they only get a writable dir in /var/tmp/, which is widely understood to be writable. But adding a new one is risky, and will piss people off. But by copying the data from trusted code, and making sure the files are owned by root, and properly chowned and with size limits applied should make this digestable for users I#d say.

**Comment by poettering** [comment]
I'd probably implement the copy logic via readdir_all(), and set the RECURSE_DIR_SORT flag there, since once we enforce an overall size limit it matters that the set of files we copy is stable if the size limit is hit.

**Comment by poettering** [comment]
I think the first and the 3rd commit are so obvious and make sense on their own that it might be worth splitting them into separate PRs and we can merge them right-away.

File: src/shared/copy.c (hunk 3 above)
Comment (AdrianVovk) [reply src/shared/copy.c:1187]: We have to check that the file is newer, not just different. Otherwise we'd only be able to sync changes to files one way.

Right now the relevant part of the sync algorithm looks something like this: Copy REPLACE_UPDATED sys -> home, then home -> sys. If we switch REPLACE_UPDATED to REPLACE_DIFFERENT, that's actually just the same thing as REPLACE. In other words, we always overwrite a file. So if you have /bulk/avatar and ~bulk/avatar, replacing the ~bulk/avatar will never do anything because the first copy step will immediately replace the ~bulk/avatar with /bulk/avatar (which has a different, but older, timestamp)

> I think we should not expect that if a home dir is portable between systems that all systems have correct clocks.

Well then that's unfortunate and someone who's using homed on systems with invalid clocks will just lose the files in their bulk directory occasionally :shrug:

File: man/systemd-homed-sync-bulk.service.xml (hunk 5 above)
Comment (AdrianVovk) [reply man/systemd-homed-sync-bulk.service.xml:16]: We want the data to be copied over continuously, on any change. Not just during activation of the home dir. Hence it's a separate service that gets triggered via a path

File: docs/USER_RECORD.md (hunk 8 above)
Comment (AdrianVovk) [reply docs/USER_RECORD.md]: I like this concept. Will rework

File: shell-completion/bash/homectl (hunk 9 above)
Comment (AdrianVovk) [reply shell-completion/bash/homectl]: At the time it was being replaced w/ avatarPath. And I'm really not convinced anything (gnome-shell, GDM, etc) will ever display the iconName

**Comment by AdrianVovk** [comment]
> I am pretty sure this should be implemented in privileged code

From privileged code, we cannot resolve the XDG subdirs

> only copy regular files, no other inode types

:+1: 

> do not descend into any dir tree

I think directories in there could be useful, for DE-specific files. To avoid namespace clashes. So if there's a GNOME wallpaper for GDM to display, it could go under /bulk/gnome/background. KDE could do their files instead in /bulk/kde/background. Thus, they won't step on each other. I suppose alternatively we could do /bulk/gnome-background or /bulk/kde-background, but that seems worse somehow :shrug:

> source should probably be placed in the top-level dir of $HOME, rather then XDG subdirs, to avoid symlink vulnerabilities.

NACK. People hate this. I hate this. I don't think anyone wants a new random top level directory or even file under ~. I understand why `~/.identity` needs to be there, but the bulk data dir could be architected to not have such a restriction

> as we copy, sum up sizes, and refuse copying more than some hardcoded limit (i.e. 64M or so)

> do not copy access mode, ownership, xattrs or such. force a fixed access mode on copied files, i.e. no x bit, no suid, world readable and stuff

> be very restrictive with the the filename charset we accept. i.e. no control chars, maybe not even non-ascii.

These all make sense to me

However, I think the point is somewhat moot overall. The problem is that the system bulk directory (the one in /var/cache) MUST be writable by external processes. Ultimately we want one place for everything to write, and it can't be the home bulk dir because that often simply doesn't exist. We can't make the synchronization one-way either, because the home directories move around and the contents of the system bulk dir might be out-of-date relative to the copy in the home directory. But the we cannot enforce that external processes follow any of these rules!

File: man/systemd-homed-sync-bulk.service.xml (hunk 5 above)
Comment (poettering) [reply man/systemd-homed-sync-bulk.service.xml:16]: do we really want that? i hate such async interfaces, since you never know when its complete. why not just make the relevant UI that drop stuff there call some bus call that syncs this down. the UI woudl then get a report when this is done. 

i.e. add a new homed bus call SynchronizeHomeBulk()?

**Comment by poettering** [comment]
> I think directories in there could be useful, for DE-specific files. To avoid namespace clashes. So if there's a GNOME wallpaper for GDM to display, it could go under /bulk/gnome/background. KDE could do their files instead in /bulk/kde/background. Thus, they won't step on each other. I suppose alternatively we could do /bulk/gnome-background or /bulk/kde-background, but that seems worse somehow 🤷

I'd really prefer a flat concept here, makes me sleep better at night. also pushes the idea a bit that this shouldn't be huge dir trees, but just some selected "bulk" files.

moreover, i have the suspicion that sooner or later there might be multiple orthogonal parameters you want to take into account when looking for a background. for example: gnome vs kde, screen size, phone vs laptop and so on. a directory tree is not that great to map that because you never know what to put at the top-level. a flat list allows more flexibility there: you can just say name your stuff background-<desktop>-<formfactor>-<resolution> and then allow them to just leave anything out they don#t care about or so.

Also it's easy to start from a list and add dirs later, it's much harder dumbing things down again

**Comment by poettering** [comment]
> NACK. People hate this. I hate this. I don't think anyone wants a new random top level directory or even file under ~. I understand why `~/.identity` needs to be there, but the bulk data dir could be architected to not have such a restriction

then you have to deal with the mess that is symlinks and the security implications they bring with them. moreover, if this is copied from priv code you don#t have access to the XDG_DATA_HOME env vars and such, which makes things nasty for crazy people who set those vars...

btw when i first thought about this i actually had a different approach in mind, i.e. a dir that is not inside $HOME, but a sibling of it. And that homed would mount below /run somewhere at the moment of activation.

That would make it an entirely different thing than $HOME, that is shipped in the same envelope however.

But I departed from the idea a bit because it only works for the LUKS backend really, for others there might be no nice place for such a sibling dir (smb....) moreover it would complicate quota for those cases too. hence i am actually fine with nesting these things inside the homedir.

**Comment by poettering** [comment]
maybe we should name the dir `~/.identity.bulk/`? so that its close relationship to `~/.identity` is communicated clearly, and people understand this is not an XDG thing, but managed from the host to some degree?

**Comment by poettering** [comment]
> However, I think the point is somewhat moot overall. The problem is that the system bulk directory (the one in /var/cache) MUST be writable by external processes.

I don't follow. Why that? What do you mean by "externel processes"? privileged ones? unpriv ones belonging to the users? why that? why can't they just request a sync?

**Comment by AdrianVovk** [comment]
> then you have to deal with the mess that is symlinks and the security implications they bring with them

Sorry I'm not groking what symlinks have to do with anything? Why can't we just exclude symlinks? I guess that would _technically_ make us out-of-spec w/ XDG (since it doesn't outright ban symlinks being part of the `XDG_*` env vars), but I can't think of any reason to point `XDG_*` at a symlink. Part of the point of these variables is that you can just set the variable instead of dealing w/ symlinks

> But I departed from the idea a bit because it only works for the LUKS backend really

Same here.

> maybe we should name the dir ~/.identity.bulk/? so that its close relationship to ~/.identity is communicated clearly, and people understand this is not an XDG thing, but managed from the host to some degree?

On a technical level sure this is OK, but on a social level I want to NACK this. Adding another random directory to ~ is very unpopular.

If we do this, I'd move the `~/.identity` file to `~/.homed/identity.json`, then have a `~/.homed/bulk/` folder next to it. To minimize the amount of files we put directly in the home dir.

Or if we want it to be a little more generic, we'd make it `~/.identity/record.json` and `~/.identity/bulk`?

> I don't follow. Why that? What do you mean by "externel processes"? privileged ones? unpriv ones belonging to the users? why that?

Part of the problem is that we don't know what, exactly, the bulk dirs are useful for. Sure: the user avatar and a background for GDM are known variables. But what else?

The example "external process" I had in mind was AccountsService. Yes, ultimately one day it would be nice to get rid of AccountsService, but that day isn't here yet. The whole GNOME stack uses AccountsService for now, for things like the profile picture of users. It's a privileged service that needs to store things like the avatar picture inside of the bulk dir.

> why can't they just request a sync?

Because the home directory might be deactivated or locked or outright gone. I expect the bulk dirs to be just a write-able as the user record when the user is deactivated.

We'd need to do extra accounting to make sure we don't try to sync w/ a deactivated or locked user. We need to be extra careful not to touch any of the files of a locked user while they are locked.

File: man/systemd-homed-sync-bulk.service.xml (hunk 5 above)
Comment (AdrianVovk) [reply man/systemd-homed-sync-bulk.service.xml:16]: Makes sense to me, but which direction is that syncing? Or is it syncing both ways?

**Comment by AdrianVovk** [comment]
Alright here's an alternate idea that you might find a bit more palatable:

**TL;DR:** Add a `bulkDirectory` field to JSON user record, to separate the concept of bulk dirs from homed. homed is the only one that can touch `/var/cache/systemd/homed/<username>.bulk/`, and we can do the security checks. If you want to change the bulk directory, you need to make a new temporary directory containing the changed files in some way and then call `Update` or some other method on homed w/ the path to the temp dir. homed takes care of the rest, and propagates changes into its own bulk directories it manages. But do read the whole proposal :)

This was motivated by me realizing that bulk directories probably shouldn't just be a homed thing, but probably also make sense in other providers of user records (remote servers, static files in /etc, and so on). Since the bulk directory is kinda part of the user record, we should define it via the user record.

So, we should add a `bulkDirectory` field to the JSON user record (regular section and per-machine). It will contain an absolute path to a world-readable bulk directory. Anything that wants a user's profile picture, for instance, should get the user record, concat the `bulkDirectory` with `/avatar`, and then access that file. Since mutating user records isn't inherently supported by the userdb, the same will be true with bulk directories. Thus, if you want to change something like the profile picture, you'll need to talk to the service that is providing the JSON user record; the same as if you were trying to change the user record itself.

Each service that provides user records will be free to synthesize `bulkDirectory` however it wants. The user record JSON files in `/etc/userdb` are free to just set `"bulkDirectory": "/path/to/some/dir"` depending on their needs, possibly even next to the record (i.e. `/etc/userdb/foobar.user` might contain `"bulkDirectory": "/etc/userdb/foobar.bulk/"`). Some remote service is free to download bulk data like profile pictures anywhere they want and then synthesize a value for `bulkDirectory`. But for homed specifically the behavior would go something like follows.

homed will manage its own bulk directories. One in `/var/cache/systemd/homed/<username>.bulk/`, which is world-readable but cannot be written by an unprivileged user. The other in the home directory itself somewhere (`~/.homed/bulk`, as I propose above?). Whenever homed returns a user record (via userdb varlink, or the various DBus APIs that return user records), it will synthesize a value for `bulkDirectory` that always points to the bulk directory in `/var/cache` if it exists. homed will never return a `bulkDirectory` other than the path under `/var/cache`.

So that covers reading bulk directories, but unlike all the other userdb providers homed also supports mutating the user records. Thus, as an extension of the user record, we need to support mutating the contents of the bulk directories. There's a couple ways I can envision this working:

1. Whenever a user record is written (i.e. via Update or Create), homed will detect if `bulkDirectory` has been set. If it has, homed will completely replace both of its managed directories with the contents of the new value (assuming the various checks you mention pass). So for example, if gnome-control-center wants to change the profile picture, it would:
    1. Create a temporary bulk directory at `/tmp/newbulk`
    2. Get the user record. Then traverse `bulkDirectory` and copy over all the files it finds into `/tmp/newbulk`, except for `/avatar` which it will be replacing.
    3. Copy the new profile picture from wherever into `/tmp/newbulk/avatar`
    4. Update the user record. It changes `bulkDirectory` to now point to `/tmp/newbulk`. Then it calls `Update` on homed with the changed user record
    5. homed replaces both of its managed bulk directories with the contents of `/tmp/newbulk`
    6. The call to `Update` returns, which means homed is done. g-c-c deletes `/tmp/newbulk`. The profile picture is changed.
2. Whenever a user record is written, if `bulkDirectory` was changed homed will try to merge the existing managed directories with the contents of the new value (again, assuming checks pass). Again, for example:
    1. Create a temporary bulk dir
    2. Copy the new profile picture from wherever into `/tmp/newbulk/avatar`
    3. Get the user record. Replace `bulkDirectory` with `/tmp/newbulk`. Then call `Update` on homed
    4. homed tries to merge the existing managed bulk directory with the contents of `/tmp/newbulk`. `/tmp/newbulk` overwrites any existing files. Any files not found in `/tmp/newbulk` are not touched in the managed directories. If g-c-c wants to delete one of the files from the bulk directory, it must make a symlink to `/dev/null` with the same filename in `/tmp/newbulk`, which indicates to homed that it should delete the file
    5. Call to `Update` returns. g-c-c deletes `/tmp/newbulk`. Done
3. We let the caller choose which way they want this to behave: merge or overwrite. When writing, `bulkDirectory` could be either a string w/ a file path or an object that takes extra fields; if it's an object, it has an extra field that takes a bool whether to merge or not. So for example:
    - If we want the first behavior we set `"bulkDirectory": "/tmp/newbulk"` or `"bulkDirectory": { "path": "/tmp/newbulk", "merge": false }`
    - If we want the second behavior, we set `"bulkDirectory": { "path": "/tmp/newbulk", "merge": true }`
4. Similar to 3, but instead of doing this in JSON we can do it in the filename? So `bulkDirectory` is always a file path, but if the file path ends in `.merge` or `.append` or something we do the merging instead of overwriting
5. We have a separate method, `UpdateBulk` or so. It could always merge. If you want to replace, you can still change `bulkDirectory` in a normal call to `Update`. This function will need to take a secret section, like `Update`. We could possibly drop the "symlink to /dev/null" stuff out of merge and just say that if you want to delete files you should call `Update` to completely replace the managed bulk dirs.
6. Similar to 5, but we'd have a boolean argument added to `UpdateBulk` that decides whether or not it merges or overwrites. Changing `bulkDirectory` during a normal `Update` will always overwrite.
7. Like 5 or 6, but we'd outright disallow changing `bulkDirectory` via `Update` and force the use of `UpdateBulk`. 

These are starting to get complicated. But each of these options have pros and cons:

1. Pro: Consistent with the behavior of `Update` regarding the records themselves. Con: Annoying to deal with from, say, the settings app. Unlike updating a record, which is just one data structure, you have to do a lot more work to copy all the files over that you don't want to get deleted.
2. Pro: Easier for the settings app. Plop in the file you want to replace/update/create/delete and go. Con: inconsistent with the behavior of `Update`. The "symlink to /dev/null" business is quite ugly, IMO.
3. Pro: Caller can choose 1 or 2 depending on what suits them best in the moment. Con: Complexity in the JSON
4. Pro: Same as 3. Con: I don't know. If we decide that we should make it possible to pick between merging and replacing, I think this is the simplest way to do it.
5. (and 6 and 7). Pro: Extra methods are more flexible API-wise. Con: Complexity.

I'm leaning most towards 1 because it's consistent, but towards 2 because it's easier on the clients of this API. Maybe 4 is the answer. I can't decide...

The last thing I want to cover in this proposal is reconciliation/sync between the two managed bulk dirs. We can do this quite easily: just look at the result of merging the records! Updating the bulk dir should bump the `lastChanged` timestamp in the user record as well. This gives is a nice and easy timestamp we can use to compare the whole directory in one go. If the embedded record is newer than the one in `/var/lib`, then we overwrite `/var/cache/...` with the bulk dir in `~`. If the system's copy of the record is newer, than we replace the home bulk dir with the one in `/var/cache/...`. That way the sync between the `/var/cache` and `~` bulk dirs corresponds perfectly to the sync of the user records

Thoughts about all this?

**Comment by AdrianVovk** [comment]
Superseded by https://github.com/systemd/systemd/pull/30840
